### PR TITLE
Draft: Added threading to AJM

### PR DESCRIPTION
**Pull request by roamic**
Very dirty code, needs a LOT of work.

### CODE REVIEW

**Comment by georgemoralis** [comment]
hint: try it on your own risk

**Comment by rafael-57** [comment]
for Bloodborne both main menu OST and menu sound effects play very consistently with this, but after going ingame and transforming weapon it crashes instantly, for some reason without an assert

it can also hang randomly during ingame load, not sure if related to these changes

File: src/core/libraries/ajm/ajm.cpp [hunk 1]
```diff
@@ -66,6 +115,91 @@ struct AjmDevice {
...
+    void WorkerThread(std::stop_token stop) {
+        while (!stop.stop_requested()) {
+            {
+                std::unique_lock lock(batch_queue_cv_mutex);
+                if (!batch_queue_cv.wait(lock, stop, [this] { return !batch_queue.empty(); })) {
+                    continue;
+                }
+            }
+
+            auto batch = batch_queue.front();
```
Comment (raphaelthegreat) [inline src/core/libraries/ajm/ajm.cpp]: This probably be inside the mutex scope

**Comment by GHU7924** [comment]
Build e7d9c5c

My run was about 2 minutes, then a crash.

[shad_log.txt](https://github.com/user-attachments/files/17562785/shad_log.txt)

Edit: I will add

https://github.com/user-attachments/assets/bdbb19be-0a73-492b-8d25-84a6c4e00727

[shad_log.txt](https://github.com/user-attachments/files/17563009/shad_log.txt)

File: src/core/libraries/ajm/ajm_instance.h [hunk 2]
```diff
@@ -12,6 +12,31 @@ struct AVCodecContext;
 struct AVCodecParserContext;
 }
 
+// Codec was not initialized properly.
+#define SCE_AJM_RESULT_NOT_INITIALIZED 0x00000001
```
Comment (raphaelthegreat) [inline src/core/libraries/ajm/ajm_instance.h]:
```suggestion
constexpr int ORBIS_AJM_RESULT_NOT_INITIALIZED = 0x00000001;
```

File: src/core/libraries/ajm/ajm.cpp [hunk 3]
```diff
@@ -66,6 +115,91 @@ struct AjmDevice {
...
+            batch_queue.pop();
+            ProcessBatch(batch->id, batch->jobs);
+            batch->finished.release();
+        }
+    }
+
+    void ProcessBatch(u32 id, std::vector<AjmJob>& jobs) {
```
Comment (raphaelthegreat) [inline src/core/libraries/ajm/ajm.cpp]:
```suggestion
    void ProcessBatch(u32 id, std::span<const AjmJob> jobs) {
```

**Comment by GHU7924** [comment]
I tried to play a few more times, but I can't pass the "ascend ladder" moment, somewhere in the middle of this ladder there is always a crash. Sometimes I manage to hear the cry of a Cleric Beast, and sometimes I don't. 

And it seems that for the first time I had an error: 

[Debug] <Critical> ajm_at9.cpp:operator():70: Assertion Failed!

[shad_log.txt](https://github.com/user-attachments/files/17574745/shad_log.txt)

**Comment by smiRaphi** [comment]
Tried it with Persona 3 Dancing Moonlight and the dlc confermation dialog now has sound :D but it crashes at the title screen with
```log
[Debug] <Critical> ajm_at9.cpp:operator():92: Assertion Failed!
Atrac9Decode failed ret = 0x82100007
```

**Comment by roamic** [comment]
Fixed one reason of crashes try again.

**Comment by squidbus** [comment]
Not sure if it was expected to be fixed or not but the Persona 3 Dancing crash is still the same.

**Comment by GHU7924** [comment]
Build a548249

"Ascend ladder" moment - FIXED, done.

https://github.com/user-attachments/assets/6950785d-e8ba-4a5d-acb7-4961a3bfe385

Then I activated the lamp, but I couldn't get into Hunter's Dream, there was no crash, but it was as if everything went into an endless loop, the loading of the location hung, but shad_log continued, I waited about a minute and closed the emulator.

Shad_log size 100 MB (therefore, I will not attach the file), but it seems like these lines were repeated

> [Lib.Ajm] <Info> ajm.cpp:sceAjmBatchStartBuffer:432: called context = 1, batch_size = 0x50, priority = 10
>[Lib.Ajm] <Info> ajm.cpp:sceAjmBatchWait:610: called context = 1, batch_id = 0, timeout = 4294967295

>[rafael-57](https://github.com/rafael-57)
>but after going ingame and transforming weapon it crashes instantly, for some reason without an assert

I couldn't test this before, but in this build, I personally don't have a crash.

https://github.com/user-attachments/assets/b16e2631-0e4b-4915-a49d-cc8eb23533c6

And another such moment:

BB

Other games have music here, but Bloodborne doesn't, I don't know if it should be like that.

Later, I will test this build again, if there is anything else, I will write.

**Comment by AwesomeSoftware79** [comment]
Build tested: [a8f0da6](https://github.com/shadps4-emu/shadPS4/actions/runs/11607165902).
CPU: Ryzen 9 7950x, Memory: DDR5 96 GB RAM, GPU: RTX 3090

BB-Patches: Skip Intro, Disable Chromatic Ab, Disable Motion Blur, 60 FPS (with Delta time), Disable vsync

In general, I would say that when it works, it just works fine. I got a couple of crashes, specially when returning to Hunter's dream and once whilst getting ambushed, when that happens it seems like there are some thread deadlocks (console spamming i.e. sceAjmBatchWait).

Some people said that this AJM Branch does not work with the 60 FPS patch applied, but my experience has been quite good. I do not see much of a difference between both 30 FPS and 60 FPS.
I tried to switch audio languages (i.e. English => Spanish) and it just works fine, i.e. NPC audio voiceovers match their configuration.

The biggest problem was with Cleric Beast's BGM, either starts looping like a broken record at some point, or does not play at all. Reloading or repeating this fight does NOT solve this problem.
Father Gascoine's fight on the other hand does present this problem. Father Gascoine's BGM changes successfully in the fight's second phase, i.e. using the [Tiny Music Box](https://bloodborne.wiki.fextralife.com/Tiny+Music+Box) three times during the fight. Gascoine's second phase BGM sounds like [this](https://www.youtube.com/watch?v=u20fQ9AdzFU).
Blood-Starved Beast loads its BGM just fine.
Audio balance between sound effects and BGM is a bit extreme. Sounds effects are quite loud and bosses' BGM is quite soft/quiet. But it could be also a problem with my audio setup.

Apart from that, at first glance I am not missing any other regular sounds.
When a issue with sound arises (apart from Cleric Beast's BGM), either reloading the save file or restarting the emulator solves the problem.

**Comment by GHU7924** [comment]
Build a8f0da6 

The first lighting of the lamp and the subsequent transfer to the Hunter's Dream did not work again, it loops (freezes) during the download. (There is a similar hanging moment in one of the videos below)
If you restart the emulator, you can already get into the Hunter's Dream, but what prevents you from doing it for the first time, I can't say.

In the Hunter's Dream, Herman's lines worked fine, 
I also knocked on the doors (4 doors in the initial location), everything is fine there, the replicas worked fine,
but for some reason Gilbert does not work properly:

https://github.com/user-attachments/assets/449bded7-acfc-4446-b447-60eda455ca3a

Then there was this:

https://github.com/user-attachments/assets/9c9c1794-f899-48fa-88dd-0360cbf72a22

The loading stopped, but the emulator continued to work.

Restarted the emulator.

https://github.com/user-attachments/assets/daf44b1a-d7df-4a4f-b302-e72db91a6b67

Next

https://github.com/user-attachments/assets/8cc70541-384c-4a78-a73e-619e91bb85d5

Next is a large gate, through which Huntsman's Minion breaks.
The sound of the gate is there, but his screams are not. (Maybe that's how it should be, I'm not sure)
But if you go a little further, you can hear the screams, but after that I had a sound loop.

https://github.com/user-attachments/assets/66d4851d-ecb7-4a73-968f-a49888509e1b

https://github.com/user-attachments/assets/2c64b62f-5a18-400a-ab3e-f57b9b684cc4

Then there was a moment when, right during the battle, the sound abruptly disappeared.
Unfortunately, I didn't record it down.

Next

https://github.com/user-attachments/assets/ac9363b9-e78a-44c4-bd23-d72f57295721

https://github.com/user-attachments/assets/86cd029a-031a-4914-9de3-8304460795fd

Perhaps something in the house affects this area, for example, an NPC, because once there was a freeze, and another time the sound disappeared.

Then I attach screenshots of the log. I don't know if they are all related to these PR or not:

Ajm
Ajm2
Ajm3
Kernel
Kernel2

In the log, it happened that the "AJM-log" simply stopped and only the "Kernel-log" continued.

I played for about an hour, I didn't visit the bosses.
That's all for now.

**Comment by GHU7924** [comment]
Before Cleric Beast (There is some kind of noise)

https://github.com/user-attachments/assets/a65d9fe3-47c3-4393-8c51-e0b3a7dd6c76

Cleric Beast (The noise remains, the musical theme is missing)

https://github.com/user-attachments/assets/e5da55a0-26a1-41df-99ab-1f0a077fc41b

Father Gascoigne (Everything seems to be in order here)

https://github.com/user-attachments/assets/02c0fe9c-9565-4993-9075-c0eb403e61f7

Father Gascoigne's daughter, The Doll, Eileen the Crow, Gehrman, The First Hunter - their lines were playing normally.

I went to Gilbert's house again to take the flamesprayer from him after defeating the Cleric Beast.

https://github.com/user-attachments/assets/76c97df5-ab64-43ef-9179-dc9ef06b1a64

Chapel Samaritan in Cathedral Ward also broken:

https://github.com/user-attachments/assets/2b417403-09ae-4289-bb0b-3acb39eca512

I didn't play anymore.

**Pull request merged**